src/stake/height_stakes.rs

```diff
@@ -0,0 +1,172 @@
+use super::Stake;
+use crate::{Error, Height};
+
+use fuel_crypto::PublicKey;
+use hashbrown::HashMap;
+
+use alloc::borrow::ToOwned;
+use core::cmp;
+use core::ops::{Bound, Range, RangeBounds};
+
+/// Mapping of a height of heights to a given stake.
+///
+/// There is a logical guarantee that the ranges will not intersect for this data structure.
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct HeightStakes {
+    keys: HashMap<Range<Height>, Stake>,
+}
+
+impl HeightStakes {
+    /// The provided key will be used to verify the consensus signatures. It might diverge from the
+    /// constant validator identifier that exists for the canonical stake contract since a key
+    /// rotation strategy is possible.
+    pub(super) fn add_stake_range<B>(&mut self, bounds: B, stake: Stake) -> Result<(), Error>
+    where
+        B: RangeBounds<Height>,
+    {
+        let bounds = normalize_range(bounds);
```

Comment:
Question for my own clarification: Is there a difference between the terms "bounds" and "range"? Or are they just used interchangeably?

Reply:
Ah - I see Rust has a formal definition: https://doc.rust-lang.org/std/ops/enum.Bound.html